// @sap/cds/libx/_runtime/common/utils/cqn2cqn4sql.js
const cds = require('../../cds')
const { SELECT, INSERT, DELETE, UPDATE } = cds.ql
const Query = require('../../../../lib/ql/Query')

const { resolveView } = require('./resolveView')
const { ensureNoDraftsSuffix } = require('./draft')
const { flattenStructuredSelect } = require('./structured')
const search2cqn4sql = require('./search2cqn4sql')
const { getEntityNameFromCQN } = require('./entityFromCqn')
const getError = require('../../common/error')
const { rewriteAsterisks } = require('./rewriteAsterisks')
const { getPathFromRef, getEntityFromPath } = require('../../common/utils/path')
const { addToWhere } = require('../../common/utils/cqn')
const { removeIsActiveEntityRecursively } = require('../../fiori/utils/where')
const { addRefToWhereIfNecessary } = require('../../../odata/afterburner')

const PARENT_ALIAS = '_parent_'
const PARENT_ALIAS_REGEX = new RegExp('^' + PARENT_ALIAS + '\\d*$')
const FOREIGN_ALIAS = '_foreign_'
const OPERATIONS = ['=', '>', '<', '!=', '<>', '>=', '<=', 'like', 'between', 'in', 'not in']

// special case in lambda functions
const _addParentAlias = (where, alias) => {
  where.forEach(e => {
    if (e.ref && e.ref[0].match(PARENT_ALIAS_REGEX)) {
      e.ref[0] = alias
    }
  })
}

const _addAliasToElement = (e, alias) => {
  if (e.ref) {
    return { ref: [alias, ...e.ref] }
  }

  if (e.list) {
    return { list: e.list.map(arg => _addAliasToElement(arg, alias)) }
  }

  if (e.func) {
    const args = e.args.map(arg => _addAliasToElement(arg, alias))
    return { ...e, args }
  }

  if (e.SELECT && e.SELECT.where) {
    _addParentAlias(e.SELECT.where, alias)
  }

  if (e.xpr) {
    return { xpr: e.xpr.map(e1 => _addAliasToElement(e1, alias)) }
  }

  return e
}

const _addAliasToExpression = (expression, alias) => {
  if (!alias) {
    return expression
  }

  return expression.map(e => _addAliasToElement(e, alias))
}

const _elementFromRef = (name, entity) => {
  if (!entity) return

  if (entity.elements) {
    return entity.elements[_getTargetFromRef(name)]
  }
}

const _getTargetFromRef = ref => {
  return ref.id || ref
}

const _getEntityName = (fromClause, entity, i) => {
  const targetName = _getTargetFromRef(fromClause.ref[i])
  return i === 0 ? targetName : entity && entity.elements[targetName] && entity.elements[targetName].target
}

const convertPathExpressionToWhere = (fromClause, model, options) => {
  if (!fromClause.ref) return
  if (fromClause.ref.length === 1) {
    const target = _getTargetFromRef(fromClause.ref[0])
    const alias = fromClause.as
    const { where, cardinality, args } = fromClause.ref[0]
    return { target, alias, where, cardinality, args }
  }

  let previousSelect, previousEntityName, previousTableAlias, structParent, previousArgs
  let prefix = []
  let columns
  for (let i = 0; i < fromClause.ref.length; i++) {
    const entity = structParent || model.definitions[previousEntityName]
    const element = _elementFromRef(fromClause.ref[i], entity)

    if (element && element._isStructured) {
      prefix.push(element.name)
      structParent = element
      continue
    } else if (element && element.isAssociation) {
      _modifyNavigationInWhere(fromClause.ref[i].where, element._target)
    } else if (element && previousSelect && i === fromClause.ref.length - 1) {
      columns = [{ ref: [...prefix, element.name] }]
      fromClause.ref.splice(-1)
      continue
    }

    const currentEntityName = _getEntityName(fromClause, entity, i)
    if (!currentEntityName) continue
    const tableAlias = `T${i}`
    const currentSelect = SELECT.from(`${currentEntityName} as ${tableAlias}`)

    if (fromClause.ref[i].where) {
      currentSelect.where(_addAliasToExpression(fromClause.ref[i].where, tableAlias))
    }

    // REVISIT: Only args in last segment are handled, intermediate ones are ignored
    const currentArgs = fromClause.ref[i].args

    if (i !== fromClause.ref.length - 1) {
      currentSelect.columns([1])
    }

    if (previousSelect) {
      const navigation = _getTargetFromRef(fromClause.ref[i])
      previousSelect.where(
        model.definitions[previousEntityName]._relations[[...prefix, navigation]].join(tableAlias, previousTableAlias)
      )
      _convertSelect(previousSelect, model, options)
      currentSelect.where('exists', previousSelect)
    }

    structParent = undefined
    prefix = []
    previousTableAlias = tableAlias
    previousSelect = currentSelect
    previousEntityName = currentEntityName
    previousArgs = currentArgs
  }

  return {
    target: previousEntityName,
    alias: previousTableAlias,
    where: previousSelect && previousSelect.SELECT && previousSelect.SELECT.where,
    args: previousArgs,
    columns
  }
}

const _convertPathExpressionForInsertOrDelete = (intoClause, model) => {
  // .into is plain string or csn entity
  if (typeof intoClause === 'string' || intoClause.name) {
    return intoClause
  }

  return intoClause.ref.reduce((res, curr, i) => {
    if (i === 0) {
      return curr.id || curr
    }

    return model.definitions[res].elements[curr.id || curr].target
  }, '')
}

const _getBottomTopRefOrVal = (func, refVal) => {
  return func.args.filter(el => el[refVal])
}

const _getWindowWhere = (where, bottomTop) => {
  const windWhere = where || []
  const bottomTopVal = _getBottomTopRefOrVal(bottomTop[0], 'val')[0]
  bottomTopVal.val = parseInt(bottomTopVal.val, 10)

  if (windWhere.length > 0) {
    windWhere.push('and')
  }

  windWhere.push({ ref: ['rowNumber'] }, '<=', bottomTopVal)
  return windWhere
}

const _getOrderByForWindowFn = bottomTop => {
  const orderBy = _getBottomTopRefOrVal(bottomTop[0], 'ref')[0]
  orderBy.sort = bottomTop[0].func === 'topcount' ? 'desc' : 'asc'
  return orderBy
}

const _getWindowXpr = (groupBy, bottomTop) => {
  const xpr = [{ func: 'ROW_NUMBER', args: [] }, 'OVER', '(']
  if (groupBy)
    xpr.push(
      'PARTITION BY',
      ...groupBy.reduce((acc, el, i) => {
        if (i < groupBy.length - 1) {
          acc.push(el)
          acc.push(',')
        } else {
          acc.push(el)
        }

        return acc
      }, [])
    )
  xpr.push('ORDER BY', _getOrderByForWindowFn(bottomTop))
  xpr.push(')')
  return { xpr: xpr, as: 'rowNumber' }
}

const _isNavCountFunc = el => el.func && el.func === 'count' && el.args[0] !== '*'

const _isCountNavigation = columns => {
  return columns.some(_isNavCountFunc)
}

const _isBottomTop = columns => {
  return columns.some(el => el.func && (el.func === 'topcount' || el.func === 'bottomcount'))
}

const _getWindColumns = (columns, groupBy, bottomTop) => {
  return [].concat(columns, _getWindowXpr(groupBy, bottomTop))
}

const _convertCountNavigation = (SELECT, model) => {
  const entityName = SELECT.from.ref[0].id || SELECT.from.ref[0]
  const entity = model.definitions[entityName]

  const newWhere = []
  for (let i = 0; i < SELECT.where.length; i++) {
    const element = SELECT.where[i]

    if (!element.func) {
      newWhere.push(element)
      continue
    }

    const navigations = element.args[0].ref
    const navigationName = navigations[0].id || navigations[0]
    if (entity.elements[navigationName]) {
      let currentEntity = entity
      let topQuery
      let lastQuery

      for (let j = 0; j < navigations.length; j++) {
        if (navigations[j].where) {
          addRefToWhereIfNecessary(navigations[j].where, currentEntity)
        }

        const nextNavigation = navigations[j]
        const nestedExistsQuery = _buildExistsQuery(
          currentEntity,
          nextNavigation,
          j > 0 ? `ALIAS_${j - 1}` : currentEntity.name,
          `ALIAS_${j}`
        )
        if (!topQuery) {
          topQuery = nestedExistsQuery
        }

        const nestedNavigationName = nextNavigation.id || nextNavigation
        if (j === navigations.length - 1) {
          nestedExistsQuery.columns('count(1) as counted')
          if (lastQuery) {
            lastQuery.where([nestedExistsQuery, SELECT.where[i + 1], SELECT.where[i + 2]])
            newWhere.push('exists', topQuery)
          } else {
            newWhere.push(nestedExistsQuery, SELECT.where[i + 1], SELECT.where[i + 2])
          }
        } else {
          if (lastQuery) {
            lastQuery.where('exists', nestedExistsQuery)
          }
        }
        lastQuery = nestedExistsQuery
        currentEntity = currentEntity.elements[nestedNavigationName]._target
      }
      i += 2
    }
  }
  SELECT.where = newWhere
}

const _addTableName = (where, tableName) => {
  return where.map(ref => {
    if (ref.ref) {
      ref.ref.unshift(tableName)
    }
    return ref
  })
}

const _buildExistsQuery = (entity, navigation, currentAlias, nextAlias) => {
  const target = entity.elements[navigation.id || navigation].target

  const existsQuery = cds.ql.SELECT.from(target)
  existsQuery.SELECT.from.as = nextAlias

  if (navigation && navigation.where) {
    existsQuery.where(_addTableName(navigation.where, nextAlias))
  }

  const onCond = entity.elements[navigation.id || navigation]._relations.join(nextAlias, currentAlias)

  return existsQuery.where(onCond)
}

const _createWindowCQN = (SELECT, model) => {
  const bottomTop = SELECT.columns.filter(el => _isBottomTop([el]))
  const columns = (SELECT.columns = [])
  const { elements } = model.definitions[SELECT.from.ref[0]]

  for (const el in elements) {
    if (!elements[el].isAssociation) {
      columns.push({ ref: [el] })
    }
  }

  SELECT.where = _getWindowWhere(SELECT.where, bottomTop)
  SELECT.from = {
    SELECT: {
      columns: _getWindColumns(columns, SELECT.groupBy, bottomTop),
      from: SELECT.from
    }
  }

  delete SELECT.groupBy
}

const _isAny = element => {
  return Array.isArray(element.ref) && element.ref.slice(-1)[0].id
}

const _isAll = element => {
  const last = Array.isArray(element.ref) && element.ref.slice(-1)[0]
  return last && last.id && last.where && last.where[0] === 'not' && last.where[1].xpr
}

// eslint-disable-next-line complexity
const _getWhereExistsSubSelect = (cqn, where, index, model, options) => {
  const _unshiftRefsWithNavigation = nav => el => {
    if (el.ref) return { ref: [...nav, ...el.ref] }
    if (el.xpr) return { xpr: el.xpr.map(_unshiftRefsWithNavigation(nav)) }
    return el
  }

  if (!options.lambdaIteration) options.lambdaIteration = 1
  const outerAlias =
    (cqn.SELECT.from.ref && cqn.SELECT.from.as) ||
    (cqn.SELECT.from.args && cqn.SELECT.from.args[0].ref && cqn.SELECT.from.args[0].as) ||
    PARENT_ALIAS + options.lambdaIteration
  const innerAlias = FOREIGN_ALIAS + options.lambdaIteration
  cqn.SELECT.from.as = outerAlias
  const ref = cqn.SELECT.from.ref || (cqn.SELECT.from.args && cqn.SELECT.from.args[0].ref)
  const queryTarget = getEntityFromPath(getPathFromRef(ref), model)

  let nav = where[index].ref.map(el => (el.id ? el.id : el))
  const last = where[index].ref.slice(-1)[0]
  const navName = queryTarget.elements[nav[0]] ? nav[0] : nav[nav.length - 1]
  const navElement = queryTarget.elements[navName]

  const lastElement = nav.reduce((csn, segment) => {
    if (csn.items) return csn // arrayed not supported
    if (csn.elements) {
      const next = csn.elements[segment]
      if (!next) return csn
      if (next.target) return model.definitions[next.target]
      return next
    }
  }, queryTarget)

  if (lastElement.items) {
    // > arrayed
    throw getError(501, `Condition expression with arrayed elements is not supported`)
  }

  if (nav[0] === outerAlias) nav = nav.slice(1)

  const condition = last.where
    ? nav.length > 1
      ? last.where.map(_unshiftRefsWithNavigation(nav.slice(1)))
      : last.where
    : undefined

  const subSelect = SELECT.from({ ref: [navElement.target], as: innerAlias })
  if (condition) {
    if (subSelect.SELECT.from.as) {
      for (let i = 0; i < condition.length; i++) {
        if (
          condition[i].ref &&
          condition[i].ref.length > 1 &&
          condition[i].ref.every(r => typeof r === 'string') &&
          condition[i].ref[0] === navName
        ) {
          condition[i].ref[0] = subSelect.SELECT.from.as
        }
      }
    }
    // skip for where: [{ val: 1 }]
    if (condition.length > 1) subSelect.where(condition)
  }

  subSelect.where(queryTarget._relations[navName].join(innerAlias, outerAlias))
  if (cds.env.effective.odata.structs) {
    flattenStructuredSelect(subSelect, model)
  }
  subSelect.columns([{ val: 1 }])

  // nested where exists needs recursive conversion
  options.lambdaIteration++

  return _convertSelect(subSelect, model, options)
}

/*
 * ensures that the where exists are fully expanded and have an infix filter
 * example: where: ['exists', { ref: ['book', 'author'] }]
 *       -> where: ['exists', { id: 'book', where: ['exists', { id: 'author', where: [{ val: 1 }] }] }]
 */
const _ensureExpandedNestedWhereExists = ({ ref }, aliases) => {
  const acc = {}
  let cur = acc
  let alias
  for (const r of ref) {
    if (Array.isArray(cur)) {
      if (cur.length > 1) {
        // > element already has condition(s)
        cur.push('and', 'exists', { ref: [{}] })
        cur = cur[cur.length - 1].ref[0]
      } else {
        cur.unshift('exists')
        cur[1] = { ref: [{}] }
        cur = cur[1].ref[0]
      }
    }
    if (typeof r === 'string' && aliases.includes(r)) {
      alias = r
    } else {
      // use Object.assign to preserve pointers
      if (typeof r === 'string') Object.assign(cur, { id: r, where: [{ val: 1 }] })
      else Object.assign(cur, r)
      cur = cur.where
    }
  }
  if (alias) return { ref: [alias, acc] }
  return { ref: [acc] }
}

const convertWhereExists = (cqn, model, options) => {
  function _recurse(where) {
    where.forEach((element, index) => {
      // ensure where exists are fully expanded
      if (where && where.some(ele => ele === 'exists')) {
        where.forEach((element, index) => {
          if (element === 'exists' && where[index + 1].ref) {
            const aliases = cqn.SELECT.from.as
              ? [cqn.SELECT.from.as]
              : cqn.SELECT.from.join
              ? cqn.SELECT.from.args.map(arg => arg.as)
              : []
            where[index + 1] = _ensureExpandedNestedWhereExists(where[index + 1], aliases)
          }
        })
      }

      if (element.xpr) {
        _recurse(element.xpr) // recursing into nested {xpr}
      } else if (element === 'exists' && _isAny(where[index + 1])) {
        where[index + 1] = _getWhereExistsSubSelect(cqn, where, index + 1, model, options)
      } else if (element === 'not' && where[index + 1] === 'exists' && _isAll(where[index + 2])) {
        where[index + 2] = _getWhereExistsSubSelect(cqn, where, index + 2, model, options)
      }
    })
  }

  const where = cqn.SELECT.where
  if (where) _recurse(where)
}

const _getRefIndex = (where, index) => {
  if (
    where[index - 1].ref &&
    // REVISIT DRAFT HANDLING: cqn2cqn4sql must not be called there
    where[index - 1].ref[where[index - 1].ref.length - 1] !== 'InProcessByUser' &&
    where[index + 1].val !== undefined &&
    where[index + 1].val !== null &&
    where[index + 1].val !== false
  ) {
    return index - 1
  }
  if (
    where[index + 1].ref &&
    // REVISIT DRAFT HANDLING: cqn2cqn4sql must not be called there
    where[index + 1].ref[where[index + 1].ref.length - 1] !== 'InProcessByUser' &&
    where[index - 1].val !== undefined &&
    where[index - 1].val !== null &&
    where[index - 1].val !== false
  ) {
    return index + 1
  }
}

const _convertNotEqual = (container, partName = 'where') => {
  const where = container[partName]

  if (where) {
    let changed
    where.forEach((el, index) => {
      if (el === '!=') {
        const refIndex = _getRefIndex(where, index)
        if (refIndex !== undefined) {
          where[index - 1] = {
            xpr: [where[index - 1], el, where[index + 1], 'or', where[refIndex], '=', { val: null }]
          }
          where[index] = where[index + 1] = undefined
          changed = true
        }
      }

      if (el && el.SELECT) {
        _convertNotEqual(el.SELECT, partName)
      }
    })

    // delete undefined values
    if (changed) {
      container[partName] = where.filter(el => el)
    }
  }

  if (container.columns) {
    container.columns.forEach(col => {
      if (typeof col === 'object') {
        if (col.SELECT) {
          _convertNotEqual(col.SELECT, partName)
        } else if (col.where) {
          _convertNotEqual(col, partName)
        }
      }
    })
  }
}

const _ifOrderByOrWhereSkip = (queryTarget, ref, model) => {
  if (!queryTarget || !queryTarget.elements[ref[0]]) return
  return ref.some(el => {
    if (typeof el !== 'string') return
    const orderbyTarget = queryTarget.elements[el]._isStructured
      ? queryTarget.elements[el]
      : model.definitions[queryTarget.elements[el].target]
    if (orderbyTarget) queryTarget = orderbyTarget
    return (
      (orderbyTarget && orderbyTarget._hasPersistenceSkip) ||
      (queryTarget.elements[el] && queryTarget.elements[el].virtual)
    )
  })
}

const _skip = (queryTarget, ref, alias, model) => {
  // we remove the alias, if it exists
  const _ref = ref[0] === alias ? ref.slice(1) : ref
  if (_ref.length === 1) return queryTarget && queryTarget.elements[_ref[0]] && queryTarget.elements[_ref[0]].virtual

  return _ifOrderByOrWhereSkip(queryTarget, _ref, model)
}

const _convertOrderByIfSkip = (orderByCQN, index) => {
  orderByCQN.splice(index, 1)
}

const _convertWhereIfSkip = (whereCQN, index) => {
  whereCQN.splice(index, 1, '1 = 1')
  OPERATIONS.includes(whereCQN[index + 1]) ? whereCQN.splice(index + 1, 2) : whereCQN.splice(index - 2, 2)
}

const _convertOrderByOrWhereCQN = (orderByOrWhereCQN, target, model, alias, processFn) => {
  const queryTarget = model.definitions[ensureNoDraftsSuffix(target)]

  orderByOrWhereCQN.forEach((el, index) => {
    if (el.ref && _skip(queryTarget, el.ref, alias, model)) processFn(orderByOrWhereCQN, index)
  })
}

const _convertOrderByOrWhereIfSkip = (cqn, target, model) => {
  const alias = cqn.SELECT.from.as
  if (cqn.SELECT.orderBy && cqn.SELECT.orderBy.length > 1) {
    _convertOrderByOrWhereCQN(cqn.SELECT.orderBy, target, model, alias, _convertOrderByIfSkip)
  }

  if (cqn.SELECT.where) {
    _convertOrderByOrWhereCQN(cqn.SELECT.where, target, model, alias, _convertWhereIfSkip)
  }
}

const _convertExpand = expand => {
  expand.forEach(expandElement => {
    if (expandElement.ref && expandElement.ref[0]) {
      if (expandElement.ref[0].where) {
        expandElement.where = expandElement.ref[0].where
      }
      if (expandElement.ref[0].id) {
        expandElement.ref[0] = expandElement.ref[0].id
      }
    }
    if (expandElement.expand) {
      _convertExpand(expandElement.expand)
    }
  })
}

const _convertRefWhereInExpand = columns => {
  if (columns) {
    columns.forEach(col => {
      if (col.expand && typeof col.expand !== 'string') {
        _convertExpand(col.expand)
      }
    })
  }
}

const _flattenCQN = cqn => {
  if (Array.isArray(cqn)) cqn.forEach(_flattenCQN)
  else if (cqn) {
    if (cqn.SELECT) _flattenCQN(cqn.SELECT)
    if (cqn.from) _flattenCQN(cqn.from)
    if (cqn.ref) _flattenCQN(cqn.ref)
    if (cqn.SET) _flattenCQN(cqn.SET)
    if (cqn.args) _flattenCQN(cqn.args)
    if (cqn.columns) _flattenCQN(cqn.columns)
    if (cqn.expand) _flattenCQN(cqn.expand)
    if (cqn.where) _flattenXpr(cqn.where)
    if (cqn.having) _flattenXpr(cqn.having)
  }
}

const _flattenXpr = cqn => {
  if (!Array.isArray(cqn)) {
    if (cqn.xpr) cqn = cqn.xpr
    return
  }

  let idx = cqn.findIndex(el => el.xpr)
  while (idx > -1) {
    cqn.splice(idx, 1, '(', ...cqn[idx].xpr, ')')
    idx = cqn.findIndex(el => el.xpr)
  }

  cqn.forEach(_flattenCQN)
}

const _convertPathExpression = (SELECT, model, options = {}) => {
  const prevAlias = SELECT.from.as
  for (const whereEl of SELECT.where || []) {
    if (typeof whereEl === 'object' && whereEl.SELECT) _convertPathExpression(whereEl.SELECT, model)
  }
  const conversion = convertPathExpressionToWhere(SELECT.from, model, options)
  if (!conversion) return
  const { target, alias, where, cardinality, columns, args } = conversion

  // REVISIT: It's not possible to just change the reference (i.e. SELECT.from.ref = [target])
  // as many parts of our code base still refer to SELECT.from (e.g. authorization)
  if (args) {
    SELECT.from = { ref: [{ id: target, args }] }
  } else {
    SELECT.from = { ref: [target] }
  }
  if (alias) {
    SELECT.from.as = alias
    if (SELECT.where && alias !== prevAlias) {
      SELECT.where = _addAliasToExpression(SELECT.where, alias)
    }
  }
  if (columns) {
    // TODO: use streaming as outer property
    if (options.isStreaming) SELECT.columns = columns
    else {
      if (!SELECT.columns) {
        // Okra always wants to have the key values, remove once we relax this requirement
        if (model.definitions[target] && model.definitions[target].keys) {
          SELECT.columns = Object.keys(model.definitions[target].keys)
            .filter(k => !model.definitions[target].keys[k].isAssociation)
            .map(k => ({ ref: [k] }))
        } else SELECT.columns = []
      }
      SELECT.columns.push(...columns)
    }
  }
  if (cardinality && cardinality.max === 1) {
    SELECT.one = true
  }
  // TODO: REVISIT: We need to add alias to subselect in .where, .columns, .from, ... etc
  if (where) {
    if (options.isDraft) {
      addToWhere({ SELECT }, where)
    } else {
      addToWhere({ SELECT }, removeIsActiveEntityRecursively(where))
    }
  }
}

// eslint-disable-next-line complexity
const _convertSelect = (query, model, _options) => {
  const options = Object.assign(
    {
      isDB: _options.service instanceof cds.DatabaseService,
      isDraft: _options.draft,
      isStreaming: query._streaming
    },
    _options
  )
  // ensure query is ql enabled
  if (!(query instanceof Query)) Object.setPrototypeOf(query, Object.getPrototypeOf(SELECT()))
  if (query.SELECT.from && query.SELECT.from.SELECT) {
    if (query.SELECT._4odata) query.SELECT.from.SELECT._4odata = true
    query.SELECT.from = _convertSelect(query.SELECT.from, model, options)
  }

  // REVISIT: a temporary workaround for xpr from new parser
  if (cds.env.features.odata_new_parser) _flattenCQN(query)

  // lambda functions
  convertWhereExists(query, model, options)

  // add 'or is null' in case of '!='
  if (query.SELECT._4odata) {
    _convertNotEqual(query.SELECT, 'where')
    _convertNotEqual(query.SELECT, 'having')
  }

  _convertPathExpression(query.SELECT, model, options)
  rewriteAsterisks(query, model, options.isDB, options.isDraft)
  if (query.SELECT.where && _isCountNavigation(query.SELECT.where)) {
    _convertCountNavigation(query.SELECT, model)
  }

  // extract where clause if it is in column expand ref
  _convertRefWhereInExpand(query.SELECT.columns)

  // REVISIT: The following operations only work for _one_ entity.
  // We must also enable them for joins etc.
  const { entityName, alias } = getEntityNameFromCQN(query)

  // remove virtual and with skip annotated fields in orderby and where
  _convertOrderByOrWhereIfSkip(query, entityName, model)

  if (query.SELECT.search && !options.suppressSearch) {
    search2cqn4sql(query, model, { ...query._searchOptions, ...{ entityName, alias } })
  }

  if (query.SELECT.columns && cds.env.effective.odata.structs) {
    flattenStructuredSelect(query, model)
  }

  // topcount with groupby
  if (query.SELECT.columns && _isBottomTop(query.SELECT.columns)) {
    _createWindowCQN(query.SELECT, model)
  }

  return query
}

const _convertInsert = (query, model, options) => {
  // resolve path expression
  const resolvedIntoClause = _convertPathExpressionForInsertOrDelete(query.INSERT.into, model)

  // overwrite only .into, foreign keys are already set
  const insert = INSERT.into(resolvedIntoClause)

  // REVISIT flatten structured types, currently its done in SQL builder

  // We add all previous properties ot the newly created query.
  // Reason is to not lose the query API functionality
  Object.assign(insert.INSERT, query.INSERT, { into: resolvedIntoClause })

  const targetName = insert.INSERT.into.name || insert.INSERT.into

  const target = model.definitions[targetName]
  if (!target) return insert

  const resolvedView = resolveView(insert, model, cds.db)

  if (query.INSERT.into.ref && query.INSERT.into.ref.length > 1) {
    const copyFrom = [...query.INSERT.into.ref]
    copyFrom.pop()
    resolvedView._validationQuery = _convertSelect(SELECT.from({ ref: copyFrom }).columns([1]), model, options)
  }

  return resolvedView
}

function _modifyNavigationInWhere(whereClause, target) {
  if (!whereClause) return
  whereClause.forEach(e => {
    if (e.ref && e.ref.length > 1 && target.elements[e.ref[0]]) {
      const element = target.elements[e.ref[0]]
      if (!element.isAssociation) return
      const foreignKeys = element._foreignKeys
      const joined = e.ref.join('_')

      for (const { parentElement } of foreignKeys) {
        if (parentElement && parentElement.name === joined) {
          e.ref = [joined]
        }
      }
    }
  })
}

const _plainDelete = (cqn, model) => {
  const name = cqn.DELETE.from.name || (cqn.DELETE.from.ref && cqn.DELETE.from.ref[0]) || cqn.DELETE.from
  const target = model.definitions[name]
  if (!target) return cqn

  return resolveView(cqn, model, cds.db)
}

const _convertDelete = (query, model, options) => {
  // .from is plain string or csn entity
  if (
    typeof query.DELETE.from === 'string' ||
    query.DELETE.from.name ||
    (query.DELETE.from.ref && typeof query.DELETE.from.ref[0] === 'string')
  ) {
    return _plainDelete(query, model)
  }

  const { target, alias, where } = convertPathExpressionToWhere(query.DELETE.from, model, options)
  const deleet = DELETE('x')
  Object.assign(deleet.DELETE, query.DELETE, { from: target, where: undefined })

  if (alias) deleet.DELETE.from = { ref: [target], as: alias }
  if (where) deleet.where(where)
  if (query.DELETE.where) deleet.where(_addAliasToExpression(query.DELETE.where, alias))

  const targetEntity = model.definitions[target]
  if (!targetEntity) return deleet

  return resolveView(deleet, model, cds.db)
}

function _plainUpdate(cqn, model) {
  const name = cqn.UPDATE.entity.name || (cqn.UPDATE.entity.ref && cqn.UPDATE.entity.ref[0]) || cqn.UPDATE.entity
  const target = model.definitions[name]
  if (!target) return cqn

  return resolveView(cqn, model, cds.db)
}

const _convertUpdate = (query, model, options) => {
  // REVISIT flatten structured types, currently its done in SQL builder

  // .into is plain string or csn entity
  if (
    typeof query.UPDATE.entity === 'string' ||
    query.UPDATE.entity.name ||
    (query.UPDATE.entity.ref && typeof query.UPDATE.entity.ref[0] === 'string' && query.UPDATE.entity.ref.length === 1)
  ) {
    return _plainUpdate(query, model)
  }

  const { target, alias, where } = convertPathExpressionToWhere(query.UPDATE.entity, model, options)

  // link .with and .data and set query target and remove current where clause
  // REVISIT: update statement does not accept cqn partial as input
  const update = UPDATE('x')
  Object.assign(update.UPDATE, query.UPDATE, { entity: target, where: undefined })

  if (alias) update.UPDATE.entity = { ref: [target], as: alias }
  if (where) update.where(where)
  if (query.UPDATE.where) update.where(_addAliasToExpression(query.UPDATE.where, alias))

  const targetEntity = model.definitions[target]
  if (!targetEntity) return update

  return resolveView(update, model, cds.db)
}

/**
 * Converts a CQN with path expression into exists clause.
 * Converts insert/update/delete on view to target table including renaming of properties
 * REVISIT structured
 * REVISIT topcount when the additional layer for Analytics before SQLBuilder is ready
 *
 * @param {object} query - incoming query
 * @param {object} model - csn model
 * @param {import('../../types/api').cqn2cqn4sqlOptions} [options] Additional options.
 */
const cqn2cqn4sql = (query, model, options = { suppressSearch: false }) => {
  if (query.SELECT) {
    return _convertSelect(query, model, options)
  }

  if (query.UPDATE) {
    return _convertUpdate(query, model, options)
  }

  if (query.INSERT) {
    return _convertInsert(query, model, options)
  }

  if (query.DELETE) {
    return _convertDelete(query, model, options)
  }

  return query
}

module.exports = {
  cqn2cqn4sql,
  convertWhereExists,
  convertPathExpressionToWhere
}
